### ISSUE TITLE
Connected component childrens not re-rendering

### ISSUE THREAD

**Issue by nlucero** [didierfranc/react-waterfall#46]
So I've started using this library recently and I'm facing an issue. I have a component that is a `Layout`: It just contains markup and renders it's children inside:

```
const Layout = ({ children }) => {
  return (
    <div>
      {/* Some fancy HTML layout */ }
      <div className="main-container">{children}</div>
    </div>
  );
}
```

Now I want to connect my `Layout` component to the `store`, so I can get the current logged-in `user` stored there:

```
connect(({ user }) => ({ user }))(Layout);
```

As soon as I connect it, the application stops behaving the way you would expect it. Components inside the layout stop re-rendering when their state/props changes.

Issue reproduced [here](https://codesandbox.io/s/qz667y1koq): Clicking on the button updates the internal state of the `Counter` component but the DOM is never updated.

I guess this happens because the state of the `store` is not changing and the `Provider` HOC is returning `false` from its `shouldComponentUpdate`.

Is this behaviour intended? How can I avoid it?

**Comment by nlucero** [comment]
I noticed that in the example if you move the `Layout` component _outside_ of the `Counter` component, it works as expected:

```
const App = () => (
  <Provider>
    <Layout>
      <Counter />
    </Layout>
  </Provider>
);
```

https://codesandbox.io/s/r7mnloxmvp

🤔

**Comment by elisherer** [comment]
I think it has something to do with an issue I found and fixed #42 (you can maybe make the change locally to see if it solves your issue)

**Comment by nlucero** [comment]
Since that is already merged I upgraded to `3.0.7` but now I see a weirder issue: Children components won't render at all on a connected component.

See: https://codesandbox.io/s/z6535l734p

**Comment by elisherer** [comment]
You are right!
Since `<Consumer>` takes children as a function (our API, not React's), the `children` prop is taken.
We should change the Consumer component API to allow children to be used as well, or at least supply an alternative (when using connect).
I'm trying to come up with a solution. (great example btw)
